### ISSUE TITLE
ivy runtime error

### ISSUE THREAD

**Issue by Corfus** [cormacrelf/angular-skyhook#512]
Skyhook works perfect with the actual default angular compiler, but when you change the compiler to ivy, you get this runtime error.

It seems to be a problem with the directives and ivy.

![Capture](https://user-images.githubusercontent.com/12698546/68474439-b449c380-0225-11ea-8318-7a730396f8dc.PNG)

**Comment by cormacrelf** [comment]
https://github.com/angular/angular/issues/25813

**Comment by cormacrelf** [comment]
Ivy is meant to be backwards compatible, although I am sure there must be caveats, as nothing in JavaScript is actually that easy. In any case, I'm going to assume Ivy-related bug are on Ivy's side until convinced otherwise, and do close to zero investigation. You can leave it open in case anyone else comes across it.

**Comment by aaronmanela** [comment]
the injection context is a breaking change. however if someone wants to use your existing build they run the basic postinstall script described here on a project that depends on angular-skyhook, it *should* work:
https://next.angular.io/guide/migration-ngcc 

`"postinstall": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points"`
this should do all the conversion for anyone using the library.

That said, when you do that the other postinstall script they mention, the for use with nguniversal,
`"postinstall": "ngcc --properties main --create-ivy-entry-points"` 
fails if using @angular-skyhook/dnd-multi-backend due to https://github.com/angular/angular/issues/33697 

This and a number of other number of breaking changes  `ng update` will fix for you and leave the library in a state that is still backwards compatible.

**Comment by cormacrelf** [comment]
This is great, thank you.

**Comment by aaronmanela** [comment]
Once you pull in the dnd-multi-backend v5 update the ngcc error should go away because the rollups are better formatted. That doesn't make it natively ivy compatible, but it will make it useable. When is that branch going to be ready? Could you make a beta release off that branch for testing purposes? It would be very helpful.

**Comment by cormacrelf** [comment]
`1.4.0-rc.0`

**Comment by cormacrelf** [comment]
#507

**Comment by Corfus** [comment]
I found the problem with ivy.
In Ivy you have to inherit from another directive.
So your abstract class DndDirective doesn't work with ivy, if you adapt it to a real Directive it should be ivy compatible, see here: 

[https://next.angular.io/guide/ivy-compatibility-examples#undecorated-classes](https://next.angular.io/guide/ivy-compatibility-examples#undecorated-classes)

In my fork it is fixed: [https://github.com/Corfus/-angular-skyhook-ivy/tree/cdb/ivy-compatibility-fix](https://github.com/Corfus/-angular-skyhook-ivy/tree/cdb/ivy-compatibility-fix)

**Comment by cormacrelf** [comment]
Is that all that needed to be done?

In any case, can you make a PR? A fork is cool, but a PR actually helps land the fix.

**Comment by Corfus** [comment]
Yes with this adaptions it works without distinctions.

Of course I can make a PR, but actually I don't have the permission to push something to your Repo. So I forked it to explain the solution.

**Comment by AlbaLopez** [comment]
@cormacrelf when do you think it will be merged?

**Comment by cormacrelf** [comment]
I think I’ll wait about 2 years, during which time I will get to some of the little things I’ve always wanted to do, like cultivate a tasteful indoor-plant aesthetic, whittle a log into the shape of a realistic Christmas ham, and train as a spider milker at the local zoo.

**Comment by AlbaLopez** [comment]
> I think I’ll wait about 2 years, during which time I will get to some of the little things I’ve always wanted to do, like cultivate a tasteful indoor-plant aesthetic, whittle a log into the shape of a realistic Christmas ham, and train as a spider milker at the local zoo.

A somewhat exaggerated answer to a question without bad intention.

Good luck and follow your dreams.

**Comment by cormacrelf** [comment]
I had hoped that merging the PR immediately would let me joke about it, but apparently no. Back to being a maintenance robot it is.

**Comment by initplatform** [comment]
Did I miss the release?

**Comment by aaronmanela** [comment]
Yay angular 9 is out! finally! I'm sticking with ViewEngine for a few months before switching to Ivy, but following this thread to see how much spider milk @cormacrelf has made. I'd like to see a picture of the ham log, too.

**Comment by palpatine1991** [comment]
Any update to this? `ngcc` advice from @aaronmanela  did not help to me and I found it impossible to use the skyhook with Angular 9 so far 😞

**Comment by aaronmanela** [comment]
Did you try using 1.4.0-rc0 branch @palpatine1991 ?

Also the ngcc postinstall fixes will no longer be recommended in Angular 9.1 as they have made it better.

**Comment by palpatine1991** [comment]
Yes, I am using skyhook 1.4.0-rc0 and I have tried it even with Angular 9.1.0. Still have the same error

**Comment by astec** [comment]
@cormacrelf do you plan to update examples to Angular 9.* with Ivy enabled? The `"@angular-skyhook/core": "^1.4.0-rc.0"` does not work with Ivy enabled for me.

**Comment by acb122** [comment]
@cormacrelf  I have made a PR which fixes the library for IVY. Could you please review it, happy to take any feedback on it, you will see it is very basic fix to make the ngcc compiler happy.
https://github.com/cormacrelf/angular-skyhook/pull/571

**Comment by palpatine1991** [comment]
@acb122 is the core of your fix adding the missing exports, upgrading dependencies or disabling the IVY?

**Comment by acb122** [comment]
@palpatine1991 the core fix is the missing exports.

IVY disabled is the current recommended setting in angular libraries in angular v10 that will change.

**Comment by MarioC3** [comment]
Hey @cormacrelf, first of all, thanks for maintaining this library! 
I was wondering when will you be able to merge PR #571 made by @acb122 ?
We're having problems with Angular 9 and this library!

Thanks again!

**Comment by a7md0** [comment]
Is there any update on this? Have anyone published #571 to new NPM package

**Comment by nzbin** [comment]
@palpatine1991 @acb122 @aaronmanela @MarioC3 @a7md0 
I have published the fixed version to npm, everything is fine with angular ivy.
[@ng-dnd/core](https://www.npmjs.com/package/@ng-dnd/core)
[@ng-dnd/multi-backend](https://www.npmjs.com/package/@ng-dnd/multi-backend)
[@ng-dnd/sortable](https://www.npmjs.com/package/@ng-dnd/sortable)

**Comment by nzbin** [comment]
@palpatine1991 @acb122 @aaronmanela @MarioC3 @a7md0 
I have published the fixed version to npm, everything is fine with angular ivy.
[@ng-dnd/core](https://www.npmjs.com/package/@ng-dnd/core)
[@ng-dnd/multi-backend](https://www.npmjs.com/package/@ng-dnd/multi-backend)
[@ng-dnd/sortable](https://www.npmjs.com/package/@ng-dnd/sortable)
